### PR TITLE
fix #679 optimize the shutdown gracefully

### LINKED ISSUE #679: About shutdown gracefully
1.在这个 https://github.com/alibaba/fescar/pull/671 PR中，我已经优化了一版shutdown gracefully，但是正确的做法应该是把RpcServer中destory方法注册为我们的钩子，否则timerExecutor不能进行优雅的关闭。

2.接口RegistryService没有close()方法，我们的资源就会得不到及时的关闭

### PR DESCRIPTION
**Pull request by CoffeeLatte007**
1. fix issue:https://github.com/alibaba/fescar/issues/679 optimize the shutdown gracefully。
2. not only server shutdown gracefully, I also fix the client shutdown gracefully with springContext shutdownHook
@github-ygy

### CODE REVIEW

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package com.alibaba.fescar.core.rpc.netty;
+
+import com.alibaba.fescar.common.util.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
```
Comment (slievrly) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java]: copyright

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java [hunk 2]
```diff
@@ -151,34 +152,30 @@ public void initChannel(SocketChannel ch) {
...
+            if (initialized.get()){
+                RegistryFactory.getInstance().unregister(new InetSocketAddress(XID.getIpAddress(), XID.getPort()));
+                RegistryFactory.getInstance().close();
+                //wait a few seconds for server transport
+                TimeUnit.SECONDS.sleep(nettyServerConfig.getServerShutdownWaitTime());
+            }
 
             this.eventLoopGroupBoss.shutdownGracefully();
-            this.eventLoopGroupWorker.shutdownGracefully();
+            this.eventLoopGroupBoss.shutdownGracefully();
```
Comment (slievrly) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java]: eventLoopGroupBoss？

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * ensure the shutdownHook is singleton
+ *
+ * @author: 563868273@qq.com
```
Comment (slievrly) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java]: not contain “：”

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 4]
```diff
@@ -107,8 +107,6 @@ public void unregister(InetSocketAddress address) throws Exception {
             return;
         }
         applicationInfoManager.setInstanceStatus(InstanceInfo.InstanceStatus.DOWN);
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: @sunrui849

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * @date: 2019/3/29
+ */
+public class FescarShutdownHook extends Thread {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FescarShutdownHook.class);
+
+    private static final FescarShutdownHook fescarShutdownHook = new FescarShutdownHook("FescarShutdownHook");
```
Comment (slievrly) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java]: FESCAR_SHUTDOWN_HOOK?

**Comment by slievrly** [comment]
pr title will be as commit log，please refer to others

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * ensure the shutdownHook is singleton
+ *
+ * @author: 563868273@qq.com
+ * @date: 2019/3/29
+ */
+public class FescarShutdownHook extends Thread {
+
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/FescarShutdownHook.java]: do not use the name like fescar

File: spring/src/main/java/com/alibaba/fescar/spring/annotation/GlobalTransactionScanner.java [hunk 7]
```diff
@@ -259,4 +277,21 @@ public void afterPropertiesSet() {
     public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
         this.applicationContext = applicationContext;
     }
+
+    private static class ShutdownHookListener implements ApplicationListener {
```
Comment (github-ygy) [inline spring/src/main/java/com/alibaba/fescar/spring/annotation/GlobalTransactionScanner.java]: why not use DisposableBean ？

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+public class ShutdownHook extends Thread {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ShutdownHook.class);
+
+    private static final ShutdownHook SHUTDOWN_HOOK = new ShutdownHook("ShutdownHook");
+
+    private Set<AbstractRpcRemoting> abstractRpcRemotings = new HashSet<>();
+
+    private final AtomicBoolean destroyed= new AtomicBoolean(false);
+
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java]: left space add between destroy =

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * ensure the shutdownHook is singleton
+ *
+ * @author 563868273@qq.com
+ * @date 2019/3/29
+ */
+public class ShutdownHook extends Thread {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ShutdownHook.class);
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java]: If Logger not used ，just remove it

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+    private Set<AbstractRpcRemoting> abstractRpcRemotings = new HashSet<>();
+
+    private final AtomicBoolean destroyed= new AtomicBoolean(false);
+
+    static {
+        Runtime.getRuntime().addShutdownHook(SHUTDOWN_HOOK);
+    }
+
+    public ShutdownHook(String name) {
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java]: how about private？

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+ * @author 563868273@qq.com
+ * @date 2019/3/29
+ */
+public class ShutdownHook extends Thread {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ShutdownHook.class);
+
+    private static final ShutdownHook SHUTDOWN_HOOK = new ShutdownHook("ShutdownHook");
+
+    private Set<AbstractRpcRemoting> abstractRpcRemotings = new HashSet<>();
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/ShutdownHook.java]: if single why not static and final ？

**Pull request merged**